src/main/java/org/weakref/jmx/ManagedClass.java

```diff
@@ -0,0 +1,147 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.weakref.jmx;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+
+import java.lang.ref.WeakReference;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+
+import static org.weakref.jmx.ReflectionUtils.getAttributeName;
+import static org.weakref.jmx.ReflectionUtils.isGetter;
+
+public class ManagedClass
+{
+    private final WeakReference<Object> target;
+    private final Map<String, ManagedClass> children;
+    private final Map<String, ManagedAttribute> attributes;
+
+    private ManagedClass(Object target, Map<String, ManagedClass> children, Map<String, ManagedAttribute> attributes)
+    {
+        this.target = new WeakReference<>(target);
+        this.children = children;
+        this.attributes = attributes;
+    }
+
+    public static ManagedClass fromExportedObject(Object target)
+    {
+        Preconditions.checkNotNull(target, "target is null");
+
+        ImmutableMap.Builder<String, ManagedAttribute> attributes = ImmutableMap.builder();
+        ImmutableMap.Builder<String, ManagedClass> children = ImmutableMap.builder();
+
+        for (Map.Entry<Method, Method> entry : AnnotationUtils.findManagedMethods(target.getClass()).entrySet()) {
+            Method concreteMethod = entry.getKey();
+            Method annotatedMethod = entry.getValue();
+
+            if (isGetter(concreteMethod)) { // is it an attribute?
+
+                String attributeName = AnnotationUtils.getName(annotatedMethod);
+                String description = AnnotationUtils.getDescription(annotatedMethod);
+
+                if (attributeName == null || attributeName.equals("")) {
+                    attributeName = getAttributeName(concreteMethod);
+                }
+
+                if (AnnotationUtils.isNested(annotatedMethod) || AnnotationUtils.isFlatten(annotatedMethod)) { // is this nested?
+                    try {
+                        Object childTarget = concreteMethod.invoke(target);
+                        if (childTarget != null) {
+                            children.put(attributeName, fromExportedObject(childTarget));
+                        }
+                    }
+                    catch (ReflectiveOperationException e) {
+                        // Ignore and continue
+                    }
+                }
+
+                attributes.put(attributeName, new ManagedAttribute(concreteMethod, attributeName, description, AnnotationUtils.isFlatten(concreteMethod)));
+            }
+        }
+        return new ManagedClass(target, children.build(), attributes.build());
+    }
+
+    public boolean isTargetPresent()
```

Comment:
What's the purpose of this method? Any calls are going to suffer from potential race conditions with the GC, so it doesn't seem as useful.

Reply:
Removed, I believe I already refactored not to use this, I'll make sure of it.